joblib/_parallel_backends.py

```diff
@@ -0,0 +1,220 @@
+"""
+Backends for embarrassingly parallel code.
+"""
+
+import gc
+import os
+import warnings
+import threading
+
+from ._multiprocessing_helpers import mp
+if mp is not None:
+    from .pool import MemmapingPool
+    from multiprocessing.pool import ThreadPool
+
+# Environment variables to protect against bad situations when nesting
+JOBLIB_SPAWNED_PROCESS = "__JOBLIB_SPAWNED_PARALLEL__"
+
+# In seconds, should be big enough to hide multiprocessing dispatching
+# overhead.
+# This settings was found by running benchmarks/bench_auto_batching.py
+# with various parameters on various platforms.
+MIN_IDEAL_BATCH_DURATION = .2
+
+# Should not be too high to avoid stragglers: long jobs running alone
+# on a single worker while other workers have no work to process any more.
+MAX_IDEAL_BATCH_DURATION = 2
+
+
+class ImmediateResult(object):
+    def __init__(self, batch):
+        # Don't delay the application, to avoid keeping the input
+        # arguments in memory
+        self.results = batch()
+
+    def get(self):
+        return self.results
+
+
+###############################################################################
+class SequentialBackend(object):
+
+    def __init__(self, parallel):
+        self.parallel = parallel
+
+    def effective_n_jobs(self, n_jobs):
+        return 1
+
+    def initialize(self, n_jobs, poolargs):
+        return 1
+
+    def compute_batch_size(self):
+        return 1
+
+    def batch_completed(self, batch_size, duration):
+        pass
+
+    def terminate(self):
+        pass
+
+    def apply_async(self, func, callback=None):
+        result = ImmediateResult(func)
+
+        if callback:
+            callback(result)
+
+        return result
+
+    def get_exceptions(self):
+        return []
+
+
+###############################################################################
+class ThreadingBackend(SequentialBackend):
+
+    def effective_n_jobs(self, n_jobs):
+        """ Determine the number of jobs which are going to run in parallel """
+        if n_jobs == 0:
+            raise ValueError('n_jobs == 0 in Parallel has no meaning')
+        elif mp is None or n_jobs is None:
+            # multiprocessing is not available or disabled, fallback
+            # to sequential mode
+            return 1
+        elif n_jobs < 0:
+            n_jobs = max(mp.cpu_count() + 1 + n_jobs, 1)

```

Comment:
I'm not sure about that. Depending on the backend, n_jobs == -1 could have a different meaning. So it should be computed in the backend class you're going to use to distribute the work on.

E.g. in the YARN context, we not going to limit ourselves to the local amount of cores available.